zkevm-circuits/src/evm_circuit/op_execution/dup.rs

```diff
@@ -0,0 +1,428 @@
+                lookups: vec![],
+            }
+        };
+
+        let stack_pointer = state_curr.stack_pointer.expr();
+        let diff = num_duplicated + stack_pointer - 1025.expr();
+        // diff's maxium is 15 when stack_pointer = 1024 and num_duplicated = 16,
+        // the minimum is 0 when num_duplicated + stack_pointer = 1025 , i.e. stack_point = 1020, num_duplicated = 5
+        let stack_underflow = {
+            Constraint {
+                name: "DupGadget stack underflow",
+                selector: self.stack_underflow.expr(),
+                polys: vec![],
+                lookups: vec![Lookup::FixedLookup(
+                    FixedLookup::Range16,
+                    [diff, 0.expr(), 0.expr()],
+                )],
+            }
+        };
+
+        let out_of_gas = {
+            let (one, two, three) = (1.expr(), 2.expr(), 3.expr());
+            let (case_selector, gas_available) = &self.out_of_gas;
+            let gas_overdemand = state_curr.gas_counter.expr()
+                + GasCost::FASTEST.expr()
+                - gas_available.expr();
+            Constraint {
+                name: "DupGadget out of gas",
+                selector: case_selector.expr(),
+                polys: vec![
+                    (gas_overdemand.clone() - one)
+                        * (gas_overdemand.clone() - two)
+                        * (gas_overdemand - three),
+                ],
+                lookups: vec![],
+            }
+        };
+
+        array::IntoIter::new([
+            success,
+            stack_overflow,
+            stack_underflow,
+            out_of_gas,
+        ])
+        .map(move |mut constraint| {
+            constraint.lookups =
+                [common_lookups.clone(), constraint.lookups].concat();
+            constraint
+        })
+        .collect()
+    }
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        core_state: &mut CoreStateInstance,
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error> {
+        match execution_step.case {
+            Case::Success => {
+                self.assign_success(region, offset, core_state, execution_step)
+            }
+            Case::StackOverflow => {
+                unimplemented!()
+            }
+            Case::StackUnderflow => {
+                unimplemented!()
+            }
+            Case::OutOfGas => {
+                unimplemented!()
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<F: FieldExt> DupGadget<F> {
+    fn assign_success(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        core_state: &mut CoreStateInstance,
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error> {
+        core_state.global_counter += 2; // read + write operation
+        core_state.program_counter += 1;
+        core_state.stack_pointer -= 1;
+        core_state.gas_counter += 3;
```

Comment:
Nitpick: these numbers appear here and between line 107~120. We could define constants for them like `byte.rs`.

```
const GC_DELTA: usize = 2;
const PC_DELTA: usize = 1;
const SP_DELTA: usize = 1;
const GAS: GasCost = GasCost::FASTEST;
```